onnx/funconnx/aionnx/_op.py

```diff
@@ -0,0 +1,198 @@
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Any, Dict
+
+import numpy as np  # type: ignore
+
+from ...onnx_pb import NodeProto
+from ..op_run import OpRun, RuntimeTypeError
+
+
+class OpRunUnary(OpRun):  # pylint: disable=W0223
+    """
+    Ancestor to all unary operators in this subfolder.
+    Checks that input and output types are the same.
+    """
+
+    def __init__(self, onnx_node: NodeProto, run_params: Dict[str, Any]):
+        OpRun.__init__(self, onnx_node, run_params)
+
+    def run(self, x):  # type: ignore # pylint: disable=W0221
+        """
+        Calls method ``_run``, catches exceptions,
+        displays a longer error message.
+        Supports only unary operators.
+        """
+        self._log("-- begin %s.run(1 input)", self.__class__.__name__)
+        try:
+            res = self._run(x)
+        except TypeError as e:
+            raise TypeError(
+                f"Issues with types {', '.join(str(type(_)) for _ in [x])} "
+                f"(unary operator {self.__class__.__name__!r})."
+            ) from e
+        self._log("-- done %s.run -> %d outputs", self.__class__.__name__, len(res))
+        return res
+
+
+class OpRunUnaryNum(OpRunUnary):  # pylint: disable=W0223
+    """
+    Ancestor to all unary and numerical operators
+    in this subfolder. Checks that input and output types
+    are the same.
+    """
+
+    def __init__(self, onnx_node: NodeProto, run_params: Dict[str, Any]):
+        OpRunUnary.__init__(self, onnx_node, run_params)
+
+    def run(self, x):  # type: ignore # pylint: disable=W0221
+        """
+        Calls method ``OpRunUnary.run``, catches exceptions,
+        displays a longer error message.
+        Checks that the result is not empty.
+        """
+        res = OpRunUnary.run(self, x)
+        if len(res) == 0 or res[0] is None:
+            return res
+        if not isinstance(res[0], list) and res[0].dtype != x.dtype:
+            raise RuntimeTypeError(
+                f"Output type mismatch: input '{x.dtype}' != output '{res[0].dtype}' "
+                f"(operator {self.__class__.__name__!r})."
+            )
+        return res
+
+
+class OpRunBinary(OpRun):  # pylint: disable=W0223
+    """
+    Ancestor to all binary operators in this subfolder.
+    Checks that input and output types are the same.
+    """
+
+    def __init__(self, onnx_node: NodeProto, run_params: Dict[str, Any]):
+        OpRun.__init__(self, onnx_node, run_params)
+
+    def run(self, x, y):  # type: ignore # pylint: disable=W0221
+        """
+        Calls method ``_run``, catches exceptions,
+        displays a longer error message.
+        Supports only binary operators.
+        """
+        self._log("-- begin %s.run(2 inputs)", self.__class__.__name__)
+        if x is None or y is None:
+            raise RuntimeError(
+                f"x and y have different dtype: {type(x)} != {type(y)} ({type(self)})"
+            )
+        if x.dtype != y.dtype:
```

Comment:
Do all op implementations do this kind of type-checking? Or, is it based on case-by-case basis?

Reply:
case by case, I assume the model was verified with onnx.checker.